### DISCUSSION TITLE
How to add additional diagram type for the same model

### DISCUSSION THREAD

**Discussion by dmm9** [eclipse-glsp/glsp#902] [Q&A]
Hi, I am trying to create different diagrams based on the same model, each showing different views of the model, and open them within the same Theia instance. But I am getting some errors in the java glsp-server when I try to add more than one DiagramModule to the same ServerModule. My environment is java-emf-theia. And this is how my approach looks like, please tell me if the concept is wrong:

- On the glsp-server (Java):
I have created two different DiagramModules, including SourceModelStorage, GModelFactory and so on. Both are configured to the same ServerModule.

- On the glsp-client (Theia):
  - Backend contributions: I have created two different extensions, each based on the GLSPSocketServerContribution. So each theia extension is actually one different glsp-client and each one opens one connection to the glsp-server.
  - Frontend contributions: As both theia-extensions (widgets) should work with the same model (file), the file-extension alone cannot be used to open the widgets. Therefore, the corresponding DiagramManagers receive the diagram-type as an openerOption to differentiate which of the widgets shall be open.

When executed individually, they work as expected. The problem is when I try to get them running at the same time in the glsp-server. I have added the `OtherDiagramModule` to the ServerModule like this:

```javascript
int port = parser.parsePort();
ServerModule tasklistServerModule = new ServerModule()
    .configureDiagramModule(new TaskListDiagramModule(), new OtherDiagramModule());

GLSPServerLauncher launcher = new SocketGLSPServerLauncher(tasklistServerModule);
launcher.start("localhost", port);
```

And when I try to open the diagram in Theia, several Java exceptions like this one occur (I have shortened the log):

```bash
Feb 04, 2023 10:02:04 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNING: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "process",
  "params": {
    "action": {
      "contextId": "tool-palette",
      "editorContext": {
        "selectedElementIds": [],
        "args": {}
      },
      "requestId": "1",
      "kind": "requestContextActions"
    },
    "clientId": "otherdiagram-diagram:file:///home/user/...../mymodel.xml"
  }
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
        [...]
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: No client session has beend initialized for client id: otherdiagram-diagram:file:///home/user/...../mymodel.xml
        at org.eclipse.glsp.server.protocol.DefaultGLSPServer.process(DefaultGLSPServer.java:165)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        ... 13 more
```
Theia opens an empty widget, without the diagram, and also throws `Uncaught (in promise) Error: Internal error.`

**Comment by tortmayr** [comment]
Hi @dmm9,

as far as I know we haven't had a usecase with multiple diagram types for the same model yet. Typically we use diagramTypes to provide multiple orthogonal diagram languages that operate on different source models. So it might be possible that there are some conceptual issues at the moment that prevent your usecase from working as expected. I need to look into this to gain more insights and will come back to you.

**Reply by dmm9** [reply]
Thanks. I am looking forward to your feedback. The exception comes from [this line](https://github.com/eclipse-glsp/glsp-server/blob/fa089aa55cdb28484617313b41ba61b1e62c9186/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/protocol/DefaultGLSPServer.java#L166) in the `DefaultGLSPServer`, in case it helps.
